cmd/activator/main.go

```diff
@@ -273,16 +281,53 @@ func main() {
 			}
 		}(name, server)
 	}
-	// Exit as soon as we see a shutdown signal or one of the servers failed.
+
+	// Wait for the signal to drain.
 	select {
-	case <-ctx.Done():
+	case <-drainCh:
+		logger.Info("Received the drain signal.")
 	case err := <-errCh:
 		logger.Errorw("Failed to run HTTP server", zap.Error(err))
 	}
 
+	// The drain has started (we are now failing readiness probes).  Let the effects of this
+	// propagate so that new requests are no longer routed our way.
+	time.Sleep(30 * time.Second)
+	logger.Info("Done waiting, shutting down servers.")
+
+	// Drain outstanding requests, and stop accepting new ones.
 	for _, server := range servers {
 		server.Shutdown(context.Background())
 	}
+	logger.Info("Servers shutdown.")
+}
+
+func healthCheck(logger *zap.SugaredLogger, statSink *websocket.ManagedConnection) (func() error, <-chan struct{}) {
+	// When we get SIGTERM (sigCh closes), start failing readiness probes.
+	sigCh := signals.SetupSignalHandler()
+
+	// Some duration after our first readiness probe failure (to allow time
+	// for the network to reprogram) send the signal to drain connections.
+	drainCh := make(chan struct{})
+	once := sync.Once{}
+
+	// The error we send once we have received SIGTERM.
+	errSigTermRecevied := errors.New("received SIGTERM from kubelet.")
+
+	return func() error {
+		select {
+		case <-sigCh:
+			// Signal to start the process of draining.
+			once.Do(func() {
+				logger.Info("Received SIGTERM")
+				close(drainCh)
+			})
+			return errSigTermRecevied
+		default:
+			logger.Debug("No signal yet.")
```

Comment:
This is gonna be super chatty...

Reply:
Trust me, it's nothing compared to what we're doing for a single Revision in the backend manager stuff.  I think that's my fault for making `updateThrottleCapacity` Info vs. Debug, but it's telling that it's called so much more than I expected.

Reply:
For contrast, we're talking 200ms (5 per second?) vs. O(a few seconds).  I had to filter the throttler logs to even see it.

Reply:
Yeah, 200ms is the current probe period in the RBM. Probably we can drop it to 100ms?